File: static/cpp/Longest_nondecreasing_subsequence.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <vector>

#include "./Longest_nondecreasing_subsequence_n2.h"
#include "./Longest_nondecreasing_subsequence_nlogn.h"

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

int main(int argc, char *argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    int n;
    if (argc == 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 10000);
      n = dis(gen);
    }
    vector<int> A;
    for (int i = 0; i < n; ++i) {
      uniform_int_distribution<int> dis(0, 99999999);
      A.emplace_back(dis(gen));
    }
    cout << "n = " << n << endl;
    int ret_length = LNS_nlogn::LongestNondecreasingSubsequenceLength(A);
    int another_length = LNS_n2::LongestNondecreasingSubsequenceLength(A);
    assert(ret_length == another_length);
  }
  return 0;
}

File: static/python/longest_nondecreasing_subsequence.py
import sys
import random
import longest_nondecreasing_subsequence_n2 as LNS_nlogn
import longest_nondecreasing_subsequence_nlogn as LNS_n2


def main():
    for times in range(1000):
        if len(sys.argv) == 2:
            n = int(sys.argv[1])
        else:
            n = random.randint(1, 10000)
        A = [random.randrange(100000000) for i in range(n)]
        print('n =', n)
        ret_length = LNS_nlogn.longest_nondecreasing_subsequence_length(A)
        another_length = LNS_n2.longest_nondecreasing_subsequence_length(A)
        assert ret_length == another_length


if __name__ == '__main__':
    main()
